### ISSUE TITLE
intake-xarray works locally but not on pangeo google cloud cluster

### ISSUE THREAD

**Issue by aaronspring** [intake/intake-xarray#86]
load data from `https://crudata.uea.ac.uk/cru/data/temperature/HadCRUT.4.6.0.0.median.nc` with simplecache, see https://github.com/aaronspring/remote_climate_data/blob/a43f5e7e4f7193caeb0633d40cca2ed8127bca30/atmosphere.yaml#L5-L38 

with `simplecache`, this first downloads data and then opens the data.

```python
import intake
cat = intake.open_catalog('https://raw.githubusercontent.com/aaronspring/remote_climate_data/master/master.yaml')
cat.atmosphere.HadCRUT4.to_dask().compute()
```

- works locally with and without calling `from dask.distributed import Client; client = Client()`
- works on pangeo cloud without using a cluster, i.e. without `client = cluster.get_client()`
- but fails on pangeo cloud with a cluster, i.e. after `client = cluster.get_client()`. But still netcdf got downloaded properly to `HadCRUT4/HadCRUT.4.6.0.0.median.nc`.

**Traceback**

<details>
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
<ipython-input-44-4cb6f7b84e70> in <module>
----> 1 obs.compute()

/srv/conda/envs/notebook/lib/python3.7/site-packages/xarray/core/dataarray.py in compute(self, **kwargs)
    832         """
    833         new = self.copy(deep=False)
--> 834         return new.load(**kwargs)
    835 
    836     def persist(self, **kwargs) -> "DataArray":

/srv/conda/envs/notebook/lib/python3.7/site-packages/xarray/core/dataarray.py in load(self, **kwargs)
    806         dask.array.compute
    807         """
--> 808         ds = self._to_temp_dataset().load(**kwargs)
    809         new = self._from_temp_dataset(ds)
    810         self._variable = new._variable

/srv/conda/envs/notebook/lib/python3.7/site-packages/xarray/core/dataset.py in load(self, **kwargs)
    652 
    653             # evaluate all the dask arrays simultaneously
--> 654             evaluated_data = da.compute(*lazy_data.values(), **kwargs)
    655 
    656             for k, data in zip(lazy_data, evaluated_data):

/srv/conda/envs/notebook/lib/python3.7/site-packages/dask/base.py in compute(*args, **kwargs)
    450         postcomputes.append(x.__dask_postcompute__())
    451 
--> 452     results = schedule(dsk, keys, **kwargs)
    453     return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
    454 

/srv/conda/envs/notebook/lib/python3.7/site-packages/distributed/client.py in get(self, dsk, keys, restrictions, loose_restrictions, resources, sync, asynchronous, direct, retries, priority, fifo_timeout, actors, **kwargs)
   2723                     should_rejoin = False
   2724             try:
-> 2725                 results = self.gather(packed, asynchronous=asynchronous, direct=direct)
   2726             finally:
   2727                 for f in futures.values():

/srv/conda/envs/notebook/lib/python3.7/site-packages/distributed/client.py in gather(self, futures, errors, direct, asynchronous)
   1990                 direct=direct,
   1991                 local_worker=local_worker,
-> 1992                 asynchronous=asynchronous,
   1993             )
   1994 

/srv/conda/envs/notebook/lib/python3.7/site-packages/distributed/client.py in sync(self, func, asynchronous, callback_timeout, *args, **kwargs)
    831         else:
    832             return sync(
--> 833                 self.loop, func, *args, callback_timeout=callback_timeout, **kwargs
    834             )
    835 

/srv/conda/envs/notebook/lib/python3.7/site-packages/distributed/utils.py in sync(loop, func, callback_timeout, *args, **kwargs)
    338     if error[0]:
    339         typ, exc, tb = error[0]
--> 340         raise exc.with_traceback(tb)
    341     else:
    342         return result[0]

/srv/conda/envs/notebook/lib/python3.7/site-packages/distributed/utils.py in f()
    322             if callback_timeout is not None:
    323                 future = asyncio.wait_for(future, callback_timeout)
--> 324             result[0] = yield future
    325         except Exception as exc:
    326             error[0] = sys.exc_info()

/srv/conda/envs/notebook/lib/python3.7/site-packages/tornado/gen.py in run(self)
    733 
    734                     try:
--> 735                         value = future.result()
    736                     except Exception:
    737                         exc_info = sys.exc_info()

/srv/conda/envs/notebook/lib/python3.7/site-packages/distributed/client.py in _gather(self, futures, errors, direct, local_worker)
   1849                             exc = CancelledError(key)
   1850                         else:
-> 1851                             raise exception.with_traceback(traceback)
   1852                         raise exc
   1853                     if errors == "skip":

/srv/conda/envs/notebook/lib/python3.7/site-packages/dask/array/core.py in getter()
    100         c = a[b]
    101         if asarray:
--> 102             c = np.asarray(c)
    103     finally:
    104         if lock:

/srv/conda/envs/notebook/lib/python3.7/site-packages/numpy/core/_asarray.py in asarray()
     81 
     82     """
---> 83     return array(a, dtype, copy=False, order=order)
     84 
     85 

/srv/conda/envs/notebook/lib/python3.7/site-packages/xarray/core/indexing.py in __array__()
    493 
    494     def __array__(self, dtype=None):
--> 495         return np.asarray(self.array, dtype=dtype)
    496 
    497     def __getitem__(self, key):

/srv/conda/envs/notebook/lib/python3.7/site-packages/numpy/core/_asarray.py in asarray()
     81 
     82     """
---> 83     return array(a, dtype, copy=False, order=order)
     84 
     85 

/srv/conda/envs/notebook/lib/python3.7/site-packages/xarray/core/indexing.py in __array__()
    653 
    654     def __array__(self, dtype=None):
--> 655         return np.asarray(self.array, dtype=dtype)
    656 
    657     def __getitem__(self, key):

/srv/conda/envs/notebook/lib/python3.7/site-packages/numpy/core/_asarray.py in asarray()
     81 
     82     """
---> 83     return array(a, dtype, copy=False, order=order)
     84 
     85 

/srv/conda/envs/notebook/lib/python3.7/site-packages/xarray/core/indexing.py in __array__()
    558     def __array__(self, dtype=None):
    559         array = as_indexable(self.array)
--> 560         return np.asarray(array[self.key], dtype=None)
    561 
    562     def transpose(self, order):

/srv/conda/envs/notebook/lib/python3.7/site-packages/numpy/core/_asarray.py in asarray()
     81 
     82     """
---> 83     return array(a, dtype, copy=False, order=order)
     84 
     85 

/srv/conda/envs/notebook/lib/python3.7/site-packages/xarray/coding/variables.py in __array__()
     68 
     69     def __array__(self, dtype=None):
---> 70         return self.func(self.array)
     71 
     72     def __repr__(self):

/srv/conda/envs/notebook/lib/python3.7/site-packages/xarray/coding/variables.py in _apply_mask()
    136 ) -> np.ndarray:
    137     """Mask all matching values in a NumPy arrays."""
--> 138     data = np.asarray(data, dtype=dtype)
    139     condition = False
    140     for fv in encoded_fill_values:

/srv/conda/envs/notebook/lib/python3.7/site-packages/numpy/core/_asarray.py in asarray()
     81 
     82     """
---> 83     return array(a, dtype, copy=False, order=order)
     84 
     85 

/srv/conda/envs/notebook/lib/python3.7/site-packages/xarray/core/indexing.py in __array__()
    558     def __array__(self, dtype=None):
    559         array = as_indexable(self.array)
--> 560         return np.asarray(array[self.key], dtype=None)
    561 
    562     def transpose(self, order):

/srv/conda/envs/notebook/lib/python3.7/site-packages/xarray/backends/netCDF4_.py in __getitem__()
     71     def __getitem__(self, key):
     72         return indexing.explicit_indexing_adapter(
---> 73             key, self.shape, indexing.IndexingSupport.OUTER, self._getitem
     74         )
     75 

/srv/conda/envs/notebook/lib/python3.7/site-packages/xarray/core/indexing.py in explicit_indexing_adapter()
    843     """
    844     raw_key, numpy_indices = decompose_indexer(key, shape, indexing_support)
--> 845     result = raw_indexing_method(raw_key.tuple)
    846     if numpy_indices.tuple:
    847         # index the loaded np.ndarray

/srv/conda/envs/notebook/lib/python3.7/site-packages/xarray/backends/netCDF4_.py in _getitem()
     82         try:
     83             with self.datastore.lock:
---> 84                 original_array = self.get_array(needs_lock=False)
     85                 array = getitem(original_array, key)
     86         except IndexError:

/srv/conda/envs/notebook/lib/python3.7/site-packages/xarray/backends/netCDF4_.py in get_array()
     61 
     62     def get_array(self, needs_lock=True):
---> 63         ds = self.datastore._acquire(needs_lock)
     64         variable = ds.variables[self.variable_name]
     65         variable.set_auto_maskandscale(False)

/srv/conda/envs/notebook/lib/python3.7/site-packages/xarray/backends/netCDF4_.py in _acquire()
    359 
    360     def _acquire(self, needs_lock=True):
--> 361         with self._manager.acquire_context(needs_lock) as root:
    362             ds = _nc4_require_group(root, self._group, self._mode)
    363         return ds

/srv/conda/envs/notebook/lib/python3.7/contextlib.py in __enter__()
    110         del self.args, self.kwds, self.func
    111         try:
--> 112             return next(self.gen)
    113         except StopIteration:
    114             raise RuntimeError("generator didn't yield") from None

/srv/conda/envs/notebook/lib/python3.7/site-packages/xarray/backends/file_manager.py in acquire_context()
    185     def acquire_context(self, needs_lock=True):
    186         """Context manager for acquiring a file."""
--> 187         file, cached = self._acquire_with_cache_info(needs_lock)
    188         try:
    189             yield file

/srv/conda/envs/notebook/lib/python3.7/site-packages/xarray/backends/file_manager.py in _acquire_with_cache_info()
    203                     kwargs = kwargs.copy()
    204                     kwargs["mode"] = self._mode
--> 205                 file = self._opener(*self._args, **kwargs)
    206                 if self._mode == "w":
    207                     # ensure file doesn't get overriden when opened again

netCDF4/_netCDF4.pyx in netCDF4._netCDF4.Dataset.__init__()

netCDF4/_netCDF4.pyx in netCDF4._netCDF4._ensure_nc_success()

FileNotFoundError: [Errno 2] No such file or directory: b'/home/jovyan/HadCRUT4/HadCRUT.4.6.0.0.median.nc'
</details>

Whats causing this issue? (also unsure where this issue belongs to)
- `intake-xarray` because it tries to open `HadCRUT4/HadCRUT.4.6.0.0.median.nc` as specified in `storage_options` and returned from `fsspec.open_local`? https://github.com/intake/intake-xarray/blob/bd43095c3f418aaf6f5cda0d3fbdf108243611e7/intake_xarray/netcdf.py#L79 
- `intake` because `intake` doesnt recognize that we are in a cloud filesystem?
- `dask_gateway`?

Related thought: maybe it would useful to have a keyword specifically ignoring `simplecache::` in the url if caching isnt requested by the user just in this specific call?

**Comment by martindurant** [comment]
Yeah, I'm not sure what to do here. The use of open_local, required for data types that must be read from a local file path, means that only a path is ever handed to dask, and the workers are not aware that this path was specially made. Similarly, you are not able to load any local file from a distributed cluster.

However, if the loader takes a file object, and we pass the more familiar OpenFile objects to dask, then each worker does download whichever files it accesses, if the URL contains caching. But in this case, you might well get multiple copies of data files, and dask does not generally know which worker has which copies.

In short, caching and distributed workers will not get along; but this still doesn't solve what you can do in the case where the loader *must* have a local file path. I'm not sure this is covered anywhere, short of having a shared cluster file-system. You could of course do it with custom delayed functions, but it is tricky to insert them into the existing netCDF load graph.

**Comment by martindurant** [comment]
Is this resolved?

**Comment by aaronspring** [comment]
nope. accepted your answer as this wont work:

> In short, caching and distributed workers will not get along
